### PR TITLE
Traction control

### PR DESCRIPTION
**Pull request by Tan560**
Hey Joe, this isn't finished. I still need to add the XML parsing for the traction control parameters, but I wanted you to look at what I have implemented and see if you have any issues. I don't think I have integrated the traction control class correctly into the constructor of swerve module, but let me know what you think

### CODE REVIEW

File: src/main/cpp/chassis/SwerveModule.cpp [hunk 1]
```diff
@@ -98,6 +99,7 @@ SwerveModule::SwerveModule(string canbusname,
     ReadConstants(configfilename);
     InitDriveMotor(driveInverted);
     InitTurnMotorEncoder(turnInverted, canCoderInverted, angleOffset, attrs);
+    m_tractionController = std::make_unique<TractionControlController>(1.2, 1.0, 0.4, 145.0, m_maxSpeed);
```
Comment (witcpalek) [inline src/main/cpp/chassis/SwerveModule.cpp:102]: read from xml

File: src/main/cpp/chassis/TractionControlController.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                                                                                                            m_optimalSlipRatio(optimalSlipRatio),
+                                                                                                            m_mass(mass / 4.0),
+                                                                                                            m_maxLinearSpeed(maxLinearSpeed),
+                                                                                                            m_maxPredictedSlipRatio((maxLinearSpeed * m_robotLoopRate_Hz) / (staticCoF * m_mass * 9.81)),
+                                                                                                            m_isSlipping(false),
+                                                                                                            m_slippingDebouncer(MIN_SLIPPING_TIME, frc::Debouncer::DebounceType::kRising)
+{
+    if (dynamicCoF > staticCoF)
+    {
+        throw std::invalid_argument("Static CoF must be higher than dynamic CoF!");
```
Comment (witcpalek) [inline src/main/cpp/chassis/TractionControlController.cpp:42]: don't throw if nobody is catching

File: src/main/cpp/chassis/TractionControlController.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    double currentSlipRatio = std::abs(inertialVelocity <= INERTIAL_VELOCITY_THRESHOLD ? wheelSpeed.value() / m_maxLinearSpeed.value()
+                                                                                       : (std::abs(wheelSpeed.value()) - inertialVelocity.value()) / inertialVelocity.value());
+
+    m_isSlipping = m_slippingDebouncer.Calculate(currentSlipRatio > m_optimalSlipRatio && std::abs(wheelSpeed.value()) > m_maxLinearSpeed.value() * m_optimalSlipRatio);
+
+    double desiredAcceleration = (velocityRequest.value() - inertialVelocity.value()) / m_robotLoopRate_Hz;
+
+    double sigmoid = 1 / (1 + std::exp(-SIGMOID_K * std::clamp(2 * (currentSlipRatio - m_optimalSlipRatio) - 1, -1.0, 1.0)));
+
+    double effectiveCoF = m_isSlipping ? m_staticCoF * (1 - sigmoid) + m_dynamicCoF * sigmoid : m_staticCoF;
```
Comment (witcpalek) [inline src/main/cpp/chassis/TractionControlController.cpp:62]: can we have constants

File: src/main/cpp/chassis/driveStates/TrajectoryDrivePathPlanner.cpp [hunk 4]
```diff
@@ -36,13 +36,13 @@ TrajectoryDrivePathPlanner::TrajectoryDrivePathPlanner(RobotDrive *robotDrive) :
                                                                                  m_trajectory(),
                                                                                  m_robotDrive(robotDrive),
                                                                                  // TODO need to tune this also update radius as it is probably wrong
-                                                                                 m_longpathHolonomicController(pathplanner::PIDConstants(1.95, 0.95, 0.0), //(0.65, 0.3755, 0.0)
-                                                                                                               pathplanner::PIDConstants(3.0, 2.5, 0.0),
+                                                                                 m_longpathHolonomicController(pathplanner::PIDConstants(1.65, 0.75, 0.0), //(1.95, 0.95, 0.0)
```
Comment (witcpalek) [inline src/main/cpp/chassis/driveStates/TrajectoryDrivePathPlanner.cpp:39]: can we use constants in .h

File: src/main/cpp/chassis/driveStates/TrajectoryDrivePathPlanner.cpp [hunk 5]
```diff
@@ -36,13 +36,13 @@ TrajectoryDrivePathPlanner::TrajectoryDrivePathPlanner(RobotDrive *robotDrive) :
...
                                                                                                                robotDrive->GetChassis()->GetMaxSpeed(),
                                                                                                                units::length::inch_t(sqrt(((robotDrive->GetChassis()->GetWheelBase().to<double>() / 2.0) * (robotDrive->GetChassis()->GetWheelBase().to<double>() / 2.0) + (robotDrive->GetChassis()->GetTrack().to<double>() / 2.0) * (robotDrive->GetChassis()->GetTrack().to<double>() / 2.0)))),
                                                                                                                units::time::second_t(0.02)),
-                                                                                 m_shortpathHolonomicController(pathplanner::PIDConstants(1.95, 0.95, 0.0),
-                                                                                                                pathplanner::PIDConstants(3.0, 2.5, 0.0),
+                                                                                 m_shortpathHolonomicController(pathplanner::PIDConstants(1.65, 0.75, 0.0),
```
Comment (witcpalek) [inline src/main/cpp/chassis/driveStates/TrajectoryDrivePathPlanner.cpp:44]: constants in the .h

**Review by witcpalek — CHANGES_REQUESTED** [review]

File: src/main/cpp/chassis/TractionControlController.cpp (hunk 2 above)
Comment (Tan560) [reply src/main/cpp/chassis/TractionControlController.cpp:42]: fixed in latest push

File: src/main/cpp/chassis/TractionControlController.cpp (hunk 3 above)
Comment (Tan560) [reply src/main/cpp/chassis/TractionControlController.cpp:73]: Sorry constants for what?

File: src/main/cpp/chassis/SwerveModule.cpp (hunk 1 above)
Comment (Tan560) [reply src/main/cpp/chassis/SwerveModule.cpp:102]: James and Aden have a branch for after kettering

**Review by witcpalek — APPROVED** [review]

**Pull request merged**